js/renovation/ui/grids/data_grid_next/scrolling/virtual_scrolling.tsx

```diff
@@ -0,0 +1,216 @@
+import {
+  SetPageIndex, PageIndex, PageSize, SetLoadPageCount,
+} from '../paging/plugins';
+import {
+  AddLoadIndexToVisibleRows, CalculateVisibleRowsInViewport,
+  TopScrollingPositionValue, RowHeightValue, ItemHeightsValue,
+  ViewportSkipValue, ViewportTakeValue, CalculateViewportSkipValue,
+  CalculateViewportTakeValue, ViewportStateValue,
+  ViewportPageIndex, ViewportLoadPageCount,
+  CalculateViewportPageIndex, CalculateViewportLoadPageCount,
+} from './plugins';
+import CLASSES from '../classes';
+import { VirtualContent } from './virtual_content';
+import { getElementHeight } from '../utils';
+import {
+  DEFAULT_ROW_HEIGHT, calculateRowHeight, calculateItemHeights, getTopScrollPosition,
+  calculateViewportItemIndex, getNormalizedPageSize, calculatePageIndexByItemIndex,
+} from './utils';
+import { Row, ViewportStateType, VirtualScrollingMode } from '../types';
+
+interface StateAction {
+  type: 'scrolling' | 'paging';
+  value: number;
+}
+
+export const viewFunction = ({
+  onRowsScrollPositionChange, topScrollPosition, onRowsViewContentRender,
+  rowHeight, itemHeights, viewportState,
+}: VirtualScrolling): JSX.Element => (
+  <Fragment>
+    <ValueSetter type={SetRowsViewScrollPositionAction} value={onRowsScrollPositionChange} />
+    <ValueSetter type={SetRowsViewContentRenderAction} value={onRowsViewContentRender} />
+    <ValueSetter type={TopScrollingPositionValue} value={topScrollPosition} />
+    <ValueSetter type={RowHeightValue} value={rowHeight} />
+    <ValueSetter type={ItemHeightsValue} value={itemHeights} />
+    <ValueSetter type={ViewportStateValue} value={viewportState} />
+    <GetterExtender type={ViewportSkipValue} order={0} value={CalculateViewportSkipValue} />
+    <GetterExtender type={ViewportTakeValue} order={0} value={CalculateViewportTakeValue} />
+    <GetterExtender type={VisibleRows} order={1} value={AddLoadIndexToVisibleRows} />
+    <GetterExtender type={VisibleRows} order={2} value={CalculateVisibleRowsInViewport} />
+    <GetterExtender type={ViewportPageIndex} order={0} value={CalculateViewportPageIndex} />
+    <GetterExtender type={ViewportLoadPageCount} order={0} value={CalculateViewportLoadPageCount} />
+    <VirtualContent />
+  </Fragment>
+);
+
+@ComponentBindings()
+export class VirtualScrollingProps {
+  @OneWay()
+  mode: VirtualScrollingMode = 'virtual';
+}
+
+@Component({
+  defaultOptionRules: null,
+  view: viewFunction,
+})
+export class VirtualScrolling extends JSXComponent(VirtualScrollingProps) {
+  @Consumer(PluginsContext)
+  plugins = new Plugins();
+
+  @InternalState()
+  topScrollPosition = 0;
+
+  @InternalState()
+  rowHeight = DEFAULT_ROW_HEIGHT;
+
+  @Mutable()
+  visibleItemHeights: Record<number, number> = {};
+
+  @InternalState()
+  viewportState: ViewportStateType = 'synchronized';
+
+  @Mutable()
+  viewportPayload = { pageIndex: 0, topScrollPosition: 0 };
+
+  get itemHeights(): Record<number, number> {
+    return this.visibleItemHeights;
+  }
```

Comment:
Why using getter that just returns mutable?

Reply:
In order to share this value among other Selectors in virtual scrolling plugins.